### PR TITLE
DAOS-3842 control: Add dmg pool set-prop command

### PR DESCRIPTION
**Pull request by mjmac**
Initial support for admin modification of pool properties. This
commit includes support for setting the "reclaim" property to
one of "(disabled|lazy|time)".

### CODE REVIEW

File: src/mgmt/srv_drpc.c [hunk 1]
```diff
@@ -572,6 +572,134 @@ ds_mgmt_drpc_pool_destroy(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+	new_prop->dpp_entries[0].dpe_type = req->number;
+
+	switch (req->value_case) {
+	case MGMT__POOL_SET_PROP_REQ__VALUE_STRVAL:
+		D_ASPRINTF(out_str_val, "%s", req->strval);
+		if (out_str_val == NULL) {
+			D_GOTO(out, rc = -DER_NOMEM);
+		}
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: nit - Don't bother changing it unless you have to change something else, but in the C code the style has been to avoid curly brackets for single-line if statements.

File: src/mgmt/srv_drpc.c [hunk 2]
```diff
@@ -572,6 +572,134 @@ ds_mgmt_drpc_pool_destroy(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
+	if (result == NULL) {
+		D_ERROR("Null set pool property response\n");
+		D_GOTO(out, rc = -DER_NOMEM);
+	}
+
+	if (result->dpp_nr != 1) {
+		D_ERROR("Expected 1 property in response (got %d)\n",
+			result->dpp_nr);
+		D_GOTO(out_result, rc = -DER_INVAL);
+	}
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: Instead of doing this, I'd suggest just trying to get the entry of the type that you set (daos_prop_entry_get). A little less rigid.

File: src/mgmt/srv_drpc.c [hunk 3]
```diff
@@ -572,6 +572,134 @@ ds_mgmt_drpc_pool_destroy(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
+
+out_result:
+	daos_prop_free(result);
+
+out:
+	daos_prop_free(new_prop);
+	D_FREE(out_str_val);
+
+	resp.status = rc;
+	len         = mgmt__pool_set_prop_resp__get_packed_size(&resp);
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: nit - don't actually need to align the assignments.

File: src/mgmt/srv_drpc.c [hunk 4]
```diff
@@ -572,6 +572,134 @@ ds_mgmt_drpc_pool_destroy(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
+	if (body == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILED_MARSHAL;
+		D_ERROR("Failed to allocate drpc response body\n");
+	} else {
+		mgmt__pool_set_prop_resp__pack(&resp, body);
+		drpc_resp->body.len  = len;
+		drpc_resp->body.data = body;
+	}
+
+	mgmt__pool_set_prop_req__free_unpacked(req, NULL);
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: Memory leak - Need to free resp.strval after it's packed.

File: src/mgmt/srv_pool.c [hunk 5]
```diff
@@ -1007,3 +1007,56 @@ ds_mgmt_pool_delete_acl(uuid_t pool_uuid, const char *principal,
...
+		goto out_ranks;
+
+	res_prop = daos_prop_alloc(prop->dpp_nr);
+	for (i = 0; i < prop->dpp_nr; i++)
+		res_prop->dpp_entries[i].dpe_type =
+			prop->dpp_entries[i].dpe_type;
+
+	rc = ds_pool_svc_get_prop(pool_uuid, ranks, res_prop);
+	if (rc != 0)
+		goto out_resprop;
```
Comment (kjacque) [inline src/mgmt/srv_pool.c]: I'd probably just free the prop here in the if statement and goto out_ranks -- personally I find that more intuitive, and you don't wind up with a goto in the successful case. Up to you though.

File: src/mgmt/tests/srv_drpc_tests.c [hunk 6]
```diff
@@ -1034,6 +1055,126 @@ test_drpc_pool_list_cont_with_containers(void **state)
...
+	setup_pool_set_prop_drpc_call(&call, &req);
+
+	ds_mgmt_drpc_pool_set_prop(&call, &resp);
+
+	expect_drpc_pool_set_prop_resp_with_error(&resp, -DER_INVAL);
+
+	D_FREE(call.body.data);
+	D_FREE(resp.body.data);
+}
+
```
Comment (kjacque) [inline src/mgmt/tests/srv_drpc_tests.c]: Should test the positive (success) case as well.

File: src/mgmt/srv_drpc.c (hunk 1 above)
Comment (mjmac) [reply src/mgmt/srv_drpc.c]: Done

File: src/mgmt/srv_drpc.c (hunk 4 above)
Comment (mjmac) [reply src/mgmt/srv_drpc.c]: d'oh, thanks

File: src/mgmt/srv_pool.c (hunk 5 above)
Comment (mjmac) [reply src/mgmt/srv_pool.c]: Yep, makes sense.

File: src/mgmt/srv_drpc.c (hunk 2 above)
Comment (mjmac) [reply src/mgmt/srv_drpc.c]: Didn't know about that... Makes sense. Fixed.

File: src/mgmt/tests/srv_drpc_tests.c (hunk 6 above)
Comment (mjmac) [reply src/mgmt/tests/srv_drpc_tests.c]: Done

File: src/mgmt/srv_drpc.c [hunk 7]
```diff
@@ -697,6 +698,8 @@ void ds_mgmt_drpc_pool_set_prop(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 		drpc_resp->body.data = body;
 	}
 
+	if (req->value_case == MGMT__POOL_SET_PROP_REQ__VALUE_STRVAL)
+		D_FREE(resp.strval);
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: IMO better to just free it. The resp is zeroed initially, so if it wasn't set, it will be NULL.

File: src/mgmt/tests/mocks.c [hunk 8]
```diff
@@ -275,9 +275,32 @@ ds_mgmt_free_pool_list(struct mgmt_list_pools_one **poolsp, uint64_t len)
...
+		res_prop->dpp_entries[i].dpe_type =
+			prop->dpp_entries[i].dpe_type;
+		res_prop->dpp_entries[i].dpe_val =
+			prop->dpp_entries[i].dpe_val;
+		res_prop->dpp_entries[i].dpe_str =
+			prop->dpp_entries[i].dpe_str;
+	}
+
+	*result = res_prop;
+	return 0;
```
Comment (kjacque) [inline src/mgmt/tests/mocks.c]: A little too much internal logic for the C mocks, in my view. It is nice to have them dumbly return whatever values the test sets up, which lets us test bizarre corner cases. For example, with this mock you can't test the case where the set appears to have succeeded, but the result doesn't include the prop type you set.

File: src/mgmt/tests/mocks.c [hunk 9]
```diff
@@ -275,9 +275,32 @@ ds_mgmt_free_pool_list(struct mgmt_list_pools_one **poolsp, uint64_t len)
...
 int
 ds_mgmt_pool_set_prop(uuid_t pool_uuid, daos_prop_t *prop,
-					  daos_prop_t **result)
+		      daos_prop_t **result)
 {
-	return ds_mgmt_set_prop_return;
+	daos_prop_t *res_prop;
+	size_t i;
+
+	if (prop == NULL)
```
Comment (kjacque) [inline src/mgmt/tests/mocks.c]: I think it's totally okay to let the unit tests crash and burn in this case. Production code won't touch this, and when I get a segfault in a unit test, the first thing I do is run gdb and inspect the stack trace. Discovery of coding errors is much quicker that way than trying to figure out the source of a surprising return code.

File: src/mgmt/tests/srv_drpc_tests.c [hunk 10]
```diff
@@ -1175,6 +1175,49 @@ test_drpc_pool_set_prop_bad_uuid(void **state)
...
+static void
+test_drpc_pool_set_prop_success(void **state)
+{
+	Drpc__Call call = DRPC__CALL__INIT;
+	Drpc__Response resp = DRPC__RESPONSE__INIT;
+	Mgmt__PoolSetPropReq req = MGMT__POOL_SET_PROP_REQ__INIT;
+	int prop_number = DAOS_PROP_PO_MAX; /* no special handling */
+	int val_number = 1;
```
Comment (kjacque) [inline src/mgmt/tests/srv_drpc_tests.c]: Nit - Need to align the names

File: src/mgmt/srv_drpc.c (hunk 7 above)
Comment (mjmac) [reply src/mgmt/srv_drpc.c]: Done

File: src/mgmt/tests/srv_drpc_tests.c (hunk 10 above)
Comment (mjmac) [reply src/mgmt/tests/srv_drpc_tests.c]: Fixed.

File: src/mgmt/tests/mocks.c (hunk 9 above)
Comment (mjmac) [reply src/mgmt/tests/mocks.c]: Done.

File: src/mgmt/tests/mocks.c (hunk 8 above)
Comment (mjmac) [reply src/mgmt/tests/mocks.c]: We're going to have to revisit this set-prop stuff at some point, would prefer to improve the tests then.

File: src/mgmt/srv_drpc.c (hunk 7 above)
Comment (mjmac) [reply src/mgmt/srv_drpc.c]: Hmm. I get a segfault if I don't guard the free. Is it a problem to do it this way?

File: src/mgmt/tests/mocks.c (hunk 8 above)
Comment (kjacque) [reply src/mgmt/tests/mocks.c]: Sure. I just feel like we should set the mock up a bit more simply to start - have a ds_mgmt_set_prop_result_out that the test sets, and just copy that to the result. You should be able to use daos_prop_dup or daos_prop_copy which simplifies the logic.

File: src/mgmt/srv_drpc.c (hunk 7 above)
Comment (kjacque) [reply src/mgmt/srv_drpc.c]: Oh, in the generated code it's a union. OK, yeah, you need to guard it. Sorry for the churn.

File: src/mgmt/tests/mocks.c [hunk 11]
```diff
@@ -272,41 +272,41 @@ ds_mgmt_free_pool_list(struct mgmt_list_pools_one **poolsp, uint64_t len)
...
+	ds_mgmt_pool_set_prop_return = 0;
+	ds_mgmt_pool_set_prop_prop = NULL;
+	ds_mgmt_pool_set_prop_result = NULL;
+	ds_mgmt_pool_set_prop_result_ptr = NULL;
 }
 
 void
 mock_ds_mgmt_pool_set_prop_teardown(void)
 {
+	daos_prop_free(ds_mgmt_pool_set_prop_result);
```
Comment (kjacque) [inline src/mgmt/tests/mocks.c]: Also need to cleanup ds_mgmt_pool_set_prop_prop since daos_prop_dup allocates it.

Comment (mjmac) [reply src/mgmt/tests/mocks.c]: https://jira.hpdd.intel.com/browse/DAOS-3938
